Out of Memory issues with v9.0.2 Maven Plugin

**Describe the bug**
We've previously been using the v8 series with great success and no major problems. However, due to the Data API changes mentioned, we tried to upgrade to the latest v9 maven plugin with the new data feed mechanism.

Now we consistently see an OutOfMemoryError exception during the NVD API download process. In the v8.x maven plugin version we used to use 2gb memory for our build slaves, but even doubling that up to 4gb, we still see OOM exceptions.

With 2gb memory, we were able to download about ~10-20% of the data before OOM. With 4gb we've managed close to double that. Somehow the new approach appears to either be leaking or downloading data into memory and the memory handling strategy is clearly different to the earlier v8.x series.

**Version of dependency-check used**
v9.0.2 maven plugin

**Log file**
```
18:36:44  [INFO] Checking for updates
18:36:48  [INFO] NVD API has 231,970 records in this update
18:36:55  [INFO] Downloaded 10,000/231,970 (4%)
18:36:59  [INFO] Downloaded 20,000/231,970 (9%)
18:37:02  [INFO] Downloaded 30,000/231,970 (13%)
18:37:09  [INFO] Downloaded 40,000/231,970 (17%)
18:37:15  [INFO] Downloaded 50,000/231,970 (22%)
18:37:22  [INFO] Downloaded 60,000/231,970 (26%)
18:37:27  [INFO] Downloaded 70,000/231,970 (30%)
18:37:37  [INFO] Downloaded 80,000/231,970 (34%)
18:37:59  [INFO] Downloaded 90,000/231,970 (39%)
18:38:04  Exception in thread "httpclient-dispatch-2" java.lang.OutOfMemoryError: Java heap space
18:38:04  	at org.apache.hc.core5.util.ByteArrayBuffer.expand(ByteArrayBuffer.java:58)
18:38:04  	at org.apache.hc.core5.util.ByteArrayBuffer.append(ByteArrayBuffer.java:88)
18:38:04  	at org.apache.hc.client5.http.async.methods.SimpleAsyncEntityConsumer.data(SimpleAsyncEntityConsumer.java:62)
18:38:04  	at org.apache.hc.core5.http.nio.entity.AbstractBinDataConsumer.consume(AbstractBinDataConsumer.java:75)
18:38:04  	at org.apache.hc.core5.http.nio.support.AbstractAsyncResponseConsumer.consume(AbstractAsyncResponseConsumer.java:134)
18:38:04  	at org.apache.hc.client5.http.impl.async.HttpAsyncMainClientExec$1.consume(HttpAsyncMainClientExec.java:243)
18:38:04  	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamHandler.consumeData(ClientHttp1StreamHandler.java:255)
18:38:04  	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamDuplexer.consumeData(ClientHttp1StreamDuplexer.java:354)
18:38:04  	at org.apache.hc.core5.http.impl.nio.AbstractHttp1StreamDuplexer.onInput(AbstractHttp1StreamDuplexer.java:325)
18:38:04  	at org.apache.hc.core5.http.impl.nio.AbstractHttp1IOEventHandler.inputReady(AbstractHttp1IOEventHandler.java:64)
18:38:04  	at org.apache.hc.core5.http.impl.nio.ClientHttp1IOEventHandler.inputReady(ClientHttp1IOEventHandler.java:41)
18:38:04  	at org.apache.hc.core5.reactor.ssl.SSLIOSession.decryptData(SSLIOSession.java:600)
18:38:04  	at org.apache.hc.core5.reactor.ssl.SSLIOSession.access$200(SSLIOSession.java:74)
18:38:04  	at org.apache.hc.core5.reactor.ssl.SSLIOSession$1.inputReady(SSLIOSession.java:202)
18:38:04  	at org.apache.hc.core5.reactor.InternalDataChannel.onIOEvent(InternalDataChannel.java:142)
18:38:04  	at org.apache.hc.core5.reactor.InternalChannel.handleIOEvent(InternalChannel.java:51)
18:38:04  	at org.apache.hc.core5.reactor.SingleCoreIOReactor.processEvents(SingleCoreIOReactor.java:178)
18:38:04  	at org.apache.hc.core5.reactor.SingleCoreIOReactor.doExecute(SingleCoreIOReactor.java:127)
18:38:04  	at org.apache.hc.core5.reactor.AbstractSingleCoreIOReactor.execute(AbstractSingleCoreIOReactor.java:86)
18:38:04  	at org.apache.hc.core5.reactor.IOReactorWorker.run(IOReactorWorker.java:44)
18:38:04  	at java.lang.Thread.run(Thread.java:750)
18:38:26  Exception in thread "httpclient-dispatch-2" java.lang.OutOfMemoryError: Java heap space
18:38:26  	at org.apache.hc.core5.util.ByteArrayBuffer.expand(ByteArrayBuffer.java:58)
18:38:26  	at org.apache.hc.core5.util.ByteArrayBuffer.append(ByteArrayBuffer.java:88)
18:38:26  	at org.apache.hc.client5.http.async.methods.SimpleAsyncEntityConsumer.data(SimpleAsyncEntityConsumer.java:62)
18:38:26  	at org.apache.hc.core5.http.nio.entity.AbstractBinDataConsumer.consume(AbstractBinDataConsumer.java:75)
```

**To Reproduce**
Suggest comparing v8.x invocation of the plugin with a 2gb maximum heap size, vs the new v9.x series with the same heap. That said, we have not seen this on every one of our projects, so it may be that there are some unique dependencies of the project that trigger this. If needed, we'll see if we can share the open source dependencies of the affected project.

**Expected behavior**
Same memory characteristics as v8.x maven plugin.

**Additional context**
N/A

Hi, 

First of all. Thanks for all the hard work you've put into this amazing plugin that impacts heavily on making the vulnerabilities known quickly. Kudos 🙇 .

I've also noticed the same out of memory -error raising after moving into the latest 9.x version from the 8.x series and the situation is bit challenging as it was informed the old version stops working in about a week or so (so we're missing the possibility to rollback into a working version).


In my experience the typical go-to answer for OOM cases ("Give some more memory to it") has almost never helped when the requirement for more memory is sudden (indicating there's significant memory leak happening somewhere in the code).

Rather than concentrating on providing OOM exception traces where the memory happens to stop (which is sort of random in any larger and more complex apps), I believe one should start looking what has caused the issues before it that lead into the situation where all the available memory has been consumed.

I'm seeing this issue in GitHub with the standard GitHub Actions runners and the logs seems to indicate there's some issue on how the database connections are handled:

```java
INSERT INTO VULNERABILITY (description, v2Severity, v2ExploitabilityScore, v2ImpactScore, v2AcInsufInfo, v2ObtainAllPrivilege, v2ObtainUserPrivilege, v2ObtainOtherPrivilege, v2UserInteractionRequired, v2Score, v2AccessVector, v2AccessComplexity, v2Authentication, v2ConfidentialityImpact, v2IntegrityImpact, v2AvailabilityImpact, v2Version, v3ExploitabilityScore, v3ImpactScore, v3AttackVector, v3AttackComplexity, v3PrivilegesRequired, v3UserInteraction, v3Scope, v3ConfidentialityImpact, v3IntegrityImpact, v3AvailabilityImpact, v3BaseScore, v3BaseSeverity, v3Version, cve) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) [50000-214]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:554)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:477)
	at org.h2.message.DbException.get(DbException.java:212)
	at org.h2.mvstore.db.Store.convertMVStoreException(Store.java:173)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:687)
	at org.h2.mvstore.db.MVPrimaryIndex.add(MVPrimaryIndex.java:126)
	at org.h2.mvstore.db.MVTable.addRow(MVTable.java:519)
	at org.h2.command.dml.Insert.insertRows(Insert.java:174)
	at org.h2.command.dml.Insert.update(Insert.java:135)
	at org.h2.command.CommandContainer.executeUpdateWithGeneratedKeys(CommandContainer.java:242)
	at org.h2.command.CommandContainer.update(CommandContainer.java:163)
	at org.h2.command.Command.executeUpdate(Command.java:252)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdateInternal(JdbcPreparedStatement.java:209)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:169)
	at org.owasp.dependencycheck.data.nvdcve.H2Functions.updateVulnerability(H2Functions.java:300)
	at jdk.internal.reflect.GeneratedMethodAccessor53.invoke(Unknown Source)
	at java.base@11.0.21/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.21/java.lang.reflect.Method.invoke(Method.java:566)
	at org.h2.schema.FunctionAlias$JavaMethod.execute(FunctionAlias.java:495)
	at org.h2.schema.FunctionAlias$JavaMethod.getTableValue(FunctionAlias.java:363)
	at org.h2.expression.function.table.JavaTableFunction.getValue(JavaTableFunction.java:34)
	at org.h2.table.FunctionTable.getResult(FunctionTable.java:51)
	at org.h2.index.VirtualConstructedTableIndex.find(VirtualConstructedTableIndex.java:38)
	at org.h2.index.IndexCursor.find(IndexCursor.java:161)
	at org.h2.table.TableFilter.next(TableFilter.java:394)
	at org.h2.command.query.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1832)
	at org.h2.result.LazyResult.hasNext(LazyResult.java:78)
	at org.h2.result.FetchedResult.next(FetchedResult.java:34)
	at org.h2.command.query.Select.queryFlat(Select.java:728)
	at org.h2.command.query.Select.queryWithoutCache(Select.java:833)
	at org.h2.command.query.Query.queryWithoutCacheLazyCheck(Query.java:197)
	at org.h2.command.query.Query.query(Query.java:512)
	at org.h2.command.query.Query.query(Query.java:475)
	at org.h2.command.CommandContainer.query(CommandContainer.java:251)
	at org.h2.command.Command.executeQuery(Command.java:190)
	at org.h2.jdbc.JdbcPreparedStatement.executeQuery(JdbcPreparedStatement.java:128)
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.executeQuery(DelegatingPreparedStatement.java:123)
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.executeQuery(DelegatingPreparedStatement.java:123)
	at org.owasp.dependencycheck.data.nvdcve.CveDB.updateOrInsertVulnerability(CveDB.java:1049)
	... 7 more
Caused by: org.h2.mvstore.MVStoreException: An old transaction with the same id is still open: 1 [2.1.214/102]
	at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:1004)
	at org.h2.mvstore.tx.TransactionStore.addUndoLogRecord(TransactionStore.java:459)
	at org.h2.mvstore.tx.Transaction.log(Transaction.java:410)
	at org.h2.mvstore.tx.TxDecisionMaker.logAndDecideToPut(TxDecisionMaker.java:158)
	at org.h2.mvstore.tx.TxDecisionMaker$PutIfAbsentDecisionMaker.decide(TxDecisionMaker.java:273)
	at org.h2.mvstore.tx.TxDecisionMaker$PutIfAbsentDecisionMaker.decide(TxDecisionMaker.java:253)
	at org.h2.mvstore.MVMap$DecisionMaker.decide(MVMap.java:1705)
	at org.h2.mvstore.MVMap.operate(MVMap.java:1779)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:346)
	at org.h2.mvstore.tx.TransactionMap.putIfAbsent(TransactionMap.java:281)
	at org.h2.mvstore.db.MVPrimaryIndex.add(MVPrimaryIndex.java:113)
	... 40 more
``` 
(having a quite a long list of these in the execution logs)

Hopefully this helps the ones more familiar with the plugin codebase to find the culprit. 

p.s. I'm using the Gradle plugin version, but I think the issue is shared by both of the plugins (as there is less active issue also open in the Gradle plugin repository).

--- a/core/src/main/java/org/owasp/dependencycheck/data/update/NvdApiDataSource.java
+++ b/core/src/main/java/org/owasp/dependencycheck/data/update/NvdApiDataSource.java
@@ -17,10 +17,13 @@
  */
 package org.owasp.dependencycheck.data.update;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
 import io.github.jeremylong.openvulnerability.client.nvd.DefCveItem;
 import io.github.jeremylong.openvulnerability.client.nvd.NvdCveClient;
 import io.github.jeremylong.openvulnerability.client.nvd.NvdCveClientBuilder;
 import java.io.File;
+import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.StringReader;
 import java.net.MalformedURLException;
@@ -41,6 +44,7 @@ import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
+import java.util.zip.GZIPOutputStream;
 import org.owasp.dependencycheck.Engine;
 import org.owasp.dependencycheck.data.nvdcve.CveDB;
 import org.owasp.dependencycheck.data.nvdcve.DatabaseException;
@@ -321,14 +325,21 @@ public class NvdApiDataSource implements CachedWebDataSource {
             int ctr = 0;
             try (NvdCveClient api = builder.build()) {
                 while (api.hasNext()) {
-                    final Collection<DefCveItem> items = api.next();
+                    Collection<DefCveItem> items = api.next();
                     max = api.getTotalAvailable();
                     if (ctr == 0) {
                         LOGGER.info(String.format("NVD API has %,d records in this update", max));
                     }
                     if (items != null && !items.isEmpty()) {
-                        final Future<NvdApiProcessor> f = processingExecutorService.submit(new NvdApiProcessor(cveDb, items));
+                        final ObjectMapper objectMapper = new ObjectMapper();
+                        objectMapper.registerModule(new JavaTimeModule());
+                        final File outputFile = settings.getTempFile("nvd-data-", ".jsonarray.gz");
+                        try (FileOutputStream fos = new FileOutputStream(outputFile);
+                                GZIPOutputStream out = new GZIPOutputStream(fos);) {
+                        objectMapper.writeValue(out, items);
+                        final Future<NvdApiProcessor> f = processingExecutorService.submit(new NvdApiProcessor(cveDb, outputFile));
                         submitted.add(f);
+                        }
                         ctr += 1;
                         if ((ctr % 5) == 0) {
                             final double percent = (double) (ctr * RESULTS_PER_PAGE) / max * 100;

--- a/core/src/main/java/org/owasp/dependencycheck/data/update/nvd/api/DownloadTask.java
+++ b/core/src/main/java/org/owasp/dependencycheck/data/update/nvd/api/DownloadTask.java
@@ -17,9 +17,6 @@
  */
 package org.owasp.dependencycheck.data.update.nvd.api;
 
-import com.fasterxml.jackson.databind.ObjectMapper;
-import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
-import io.github.jeremylong.openvulnerability.client.nvd.CveApiJson20;
 import java.io.File;
 import java.net.URL;
 import java.util.concurrent.Callable;
@@ -28,7 +25,6 @@ import java.util.concurrent.Future;
 import javax.annotation.concurrent.ThreadSafe;
 import org.apache.commons.lang3.StringUtils;
 import org.owasp.dependencycheck.data.nvdcve.CveDB;
-import org.owasp.dependencycheck.data.update.exception.UpdateException;
 import org.owasp.dependencycheck.utils.Downloader;
 import org.owasp.dependencycheck.utils.Settings;
 import org.slf4j.Logger;

--- a/core/src/main/java/org/owasp/dependencycheck/data/update/nvd/api/DownloadTask.java
+++ b/core/src/main/java/org/owasp/dependencycheck/data/update/nvd/api/DownloadTask.java
@@ -20,6 +20,7 @@ package org.owasp.dependencycheck.data.update.nvd.api;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
 import io.github.jeremylong.openvulnerability.client.nvd.CveApiJson20;
+import java.io.File;
 import java.net.URL;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutorService;
@@ -72,7 +73,6 @@ public class DownloadTask implements Callable<Future<NvdApiProcessor>> {
      * @param settings a reference to the global settings object; this is
      * necessary so that when the thread is started the dependencies have a
      * correct reference to the global settings.
-     * @throws UpdateException thrown if temporary files could not be created
      */
     public DownloadTask(String url, ExecutorService processor, CveDB cveDB, Settings settings) {
         this.url = url;
@@ -89,15 +89,12 @@ public class DownloadTask implements Callable<Future<NvdApiProcessor>> {
             LOGGER.info("Download Started for NVD Cache - {}", url);
             final long startDownload = System.currentTimeMillis();
             final Downloader d = new Downloader(settings);
-            final String content = d.fetchGzContent(u, true, Settings.KEYS.NVD_API_DATAFEED_USER, Settings.KEYS.NVD_API_DATAFEED_PASSWORD);
-            final ObjectMapper objectMapper = new ObjectMapper();
-            objectMapper.registerModule(new JavaTimeModule());
-            final CveApiJson20 data = objectMapper.readValue(content, CveApiJson20.class);
-
+            final File outputFile = settings.getTempFile("nvd-datafeed-", "json.gz");
+            d.fetchFile(u, outputFile, true, Settings.KEYS.NVD_API_DATAFEED_USER, Settings.KEYS.NVD_API_DATAFEED_PASSWORD);
             if (this.processorService == null) {
                 return null;
             }
-            final NvdApiProcessor task = new NvdApiProcessor(cveDB, data.getVulnerabilities(), startDownload);
+            final NvdApiProcessor task = new NvdApiProcessor(cveDB, outputFile, startDownload);
             final Future<NvdApiProcessor> val = this.processorService.submit(task);
             return val;
         } catch (Throwable ex) {

--- a/core/src/main/java/org/owasp/dependencycheck/data/update/nvd/api/NvdApiProcessor.java
+++ b/core/src/main/java/org/owasp/dependencycheck/data/update/nvd/api/NvdApiProcessor.java
@@ -17,11 +17,21 @@
  */
 package org.owasp.dependencycheck.data.update.nvd.api;
 
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import io.github.jeremylong.openvulnerability.client.nvd.CveApiJson20;
 import io.github.jeremylong.openvulnerability.client.nvd.DefCveItem;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
 import java.util.Collection;
+import java.util.List;
 import java.util.concurrent.Callable;
+import java.util.zip.GZIPInputStream;
 import org.owasp.dependencycheck.data.nvd.ecosystem.CveEcosystemMapper;
 import org.owasp.dependencycheck.data.nvdcve.CveDB;
+import org.owasp.dependencycheck.data.update.exception.UpdateException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -41,9 +51,9 @@ public class NvdApiProcessor implements Callable<NvdApiProcessor> {
      */
     private final CveDB cveDB;
     /**
-     * The collection of NVD API data to add to the database.
+     * The file containing the data to inject.
      */
-    private Collection<DefCveItem> data;
+    private File jsonFile;
     /**
      * Reference to the CVE Ecosystem Mapper object.
      */
@@ -60,13 +70,13 @@ public class NvdApiProcessor implements Callable<NvdApiProcessor> {
     /**
      * Create a new processor to put the NVD data into the database.
      *
-     * @param cveDB a reference to the database
-     * @param data the data to add to the database
+     * @param cveDB a reference to the database.
+     * @param jsonFile the JSON data file to inject.
      * @param startTime the start time of the update process.
      */
-    public NvdApiProcessor(final CveDB cveDB, Collection<DefCveItem> data, long startTime) {
+    public NvdApiProcessor(final CveDB cveDB, File jsonFile, long startTime) {
         this.cveDB = cveDB;
-        this.data = data;
+        this.jsonFile = jsonFile;
         this.startTime = startTime;
     }
 
@@ -74,23 +84,48 @@ public class NvdApiProcessor implements Callable<NvdApiProcessor> {
      * Create a new processor to put the NVD data into the database.
      *
      * @param cveDB a reference to the database
-     * @param data the data to add to the database
+     * @param jsonFile the JSON data file to inject.
      */
-    public NvdApiProcessor(final CveDB cveDB, Collection<DefCveItem> data) {
-        this(cveDB, data, System.currentTimeMillis());
+    public NvdApiProcessor(final CveDB cveDB, File jsonFile) {
+        this(cveDB, jsonFile, System.currentTimeMillis());
     }
 
     @Override
     public NvdApiProcessor call() throws Exception {
-        for (DefCveItem entry : data) {
-            try {
-            cveDB.updateVulnerability(entry, mapper.getEcosystem(entry));
-            } catch (Exception ex) {
-                LOGGER.error("Failed to process " + entry.getCve().getId(), ex);
+        final ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.registerModule(new JavaTimeModule());
+         Collection<DefCveItem> data = null;
+        
+         if (jsonFile.getName().endsWith(".jsonarray.gz")) {
+            try (FileInputStream fileInputStream = new FileInputStream(jsonFile);
+                        GZIPInputStream gzipInputStream = new GZIPInputStream(fileInputStream);) {
+                data = objectMapper.readValue(gzipInputStream, new TypeReference<Collection<DefCveItem>>(){});
+            } catch (IOException exception) {
+                throw new UpdateException("Unable to read downloaded json data: " + jsonFile, exception);
+            }
+        } else if (jsonFile.getName().endsWith(".gz")) {
+            try (FileInputStream fileInputStream = new FileInputStream(jsonFile);
+                        GZIPInputStream gzipInputStream = new GZIPInputStream(fileInputStream);) {
+                CveApiJson20 cveData = objectMapper.readValue(gzipInputStream, CveApiJson20.class);
+                if (cveData != null) {
+                    data = cveData.getVulnerabilities();
+                }
+            } catch (IOException exception) {
+                throw new UpdateException("Unable to read downloaded json data: " + jsonFile, exception);
+            }
+        } else {
+            data = objectMapper.readValue(jsonFile, new TypeReference<Collection<DefCveItem>>(){});
+        }
+        if (data != null ) {
+            for (DefCveItem entry : data) {
+                try {
+                    cveDB.updateVulnerability(entry, mapper.getEcosystem(entry));
+                } catch (Exception ex) {
+                    LOGGER.error("Failed to process " + entry.getCve().getId(), ex);
+                }
             }
         }
         endTime = System.currentTimeMillis();
-        data = null;
         return this;
     }
 

--- a/core/src/main/java/org/owasp/dependencycheck/data/update/nvd/api/NvdApiProcessor.java
+++ b/core/src/main/java/org/owasp/dependencycheck/data/update/nvd/api/NvdApiProcessor.java
@@ -26,7 +26,6 @@ import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.util.Collection;
-import java.util.List;
 import java.util.concurrent.Callable;
 import java.util.zip.GZIPInputStream;
 import org.owasp.dependencycheck.data.nvd.ecosystem.CveEcosystemMapper;

--- a/utils/src/main/java/org/owasp/dependencycheck/utils/Downloader.java
+++ b/utils/src/main/java/org/owasp/dependencycheck/utils/Downloader.java
@@ -178,49 +178,11 @@ public final class Downloader {
     public String fetchContent(URL url, boolean useProxy, String userKey, String passwordKey)
             throws DownloadFailedException, TooManyRequestsException, ResourceNotFoundException {
         InputStream in = null;
-        try (HttpResourceConnection conn = new HttpResourceConnection(settings, useProxy, userKey, passwordKey);
+        try (HttpResourceConnection conn = new HttpResourceConnection(settings, useProxy, userKey, passwordKey); 
                 ByteArrayOutputStream out = new ByteArrayOutputStream()) {
             in = conn.fetch(url);
             IOUtils.copy(in, out);
             return out.toString(UTF8);
-        } catch (IOException ex) {
-            final String msg = format("Download failed, unable to retrieve '%s'; %s", url, ex.getMessage());
-            throw new DownloadFailedException(msg, ex);
-        } finally {
-            if (in != null) {
-                try {
-                    in.close();
-                } catch (IOException ex) {
-                    LOGGER.trace("Ignorable error", ex);
-                }
-            }
-        }
-    }
-
-    /**
-     * Retrieves a gzip file from a given URL and returns the uncompressed contents.
-     *
-     * @param url the URL of the file to download
-     * @param useProxy whether to use the configured proxy when downloading
-     * files
-     * @return the content of the file
-     * @param userKey the settings key for the username to be used
-     * @param passwordKey the settings key for the password to be used
-     * @throws DownloadFailedException is thrown if there is an error
-     * downloading the file
-     * @throws TooManyRequestsException thrown when a 429 is received
-     * @throws ResourceNotFoundException thrown when a 404 is received
-     */
-    public String fetchGzContent(URL url, boolean useProxy, String userKey, String passwordKey)
-            throws DownloadFailedException, TooManyRequestsException, ResourceNotFoundException {
-        InputStream in = null;
-        try (HttpResourceConnection conn = new HttpResourceConnection(settings, useProxy, userKey, passwordKey)) {
-            in = conn.fetch(url);
-            try (GZIPInputStream gzipIn = new GZIPInputStream(in);
-                    ByteArrayOutputStream out = new ByteArrayOutputStream()) {
-                IOUtils.copy(gzipIn, out);
-                return out.toString(UTF8);
-            }
         } catch (IOException ex) {
             final String msg = format("Download failed, unable to retrieve '%s'; %s", url, ex.getMessage());
             throw new DownloadFailedException(msg, ex);

